### PR TITLE
Investment project stages

### PR DESCRIPTION
**Pull request by feedmypixel**
**NOTE - the Investment Project flow itself is in discussion at present as we need some discovery to confirm we are meeting user needs. However the work here can be easily repurposed and is still valid**

This work involves the first part of the investment creation project flow. 

### Of Note (see commits for more detailed information)
- Rework investment project flow to work with CTI, FDI and Non-FDI types (see flow diagram below)
- Improve url structure for the create flow
- Add new form macros to Investment project creation flow
- Add project stage progress component to show current position in the flow, #423 
- Add move project stage component to move between stages, #422 
- Add middleware for storing information under a namespace in the session
- Add date component pattern #425 

### TODO
- There are a number of UX issues to address, this will be addressed in the next piece of work.
   - what you need to do to proceed to the next stage
   - position of stage info/message box
   - Requirements and Value forms asking to much
   - forms too big
- Please note stronger test coverage will be coming in a future PR.

### Create Investment project journey flow diagram
create-investment-porject-journey-flow

### Investment project journey (uk and foreign)
investment-project

### CODE REVIEW

File: assets/stylesheets/components/_progress.scss [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    background-color: $grey-3;
+    z-index: 10;
+  }
+
+  &::after {
+    background-color: $govuk-blue;
+    z-index: 20;
+  }
+
+  &.c-progress__bar--prospect-stage {
```
Comment (teneightfive) [inline assets/stylesheets/components/_progress.scss]: I'd be tempted not to use class names that relate to investment progress only. Maybe modifiers that relate to the percentage complete?

File: assets/stylesheets/components/_progress.scss [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+.c-progress__stage--reset {
+  &::before {
+    background-color: $white;
+    border: 2px solid $grey-3;
+  }
+}
+
+.c-progress__stage--active {
```
Comment (teneightfive) [inline assets/stylesheets/components/_progress.scss]: This feels more like a state class than a modifier. Maybe just `is-active`.

File: assets/stylesheets/components/_progress.scss [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+.c-progress__stage--part-4 {
+  left: $section-width * 3;
+}
+
+.c-progress__stage--part-5 {
+  left: $section-width * 4;
+}
+
+@include media($min-width: 1060px) {
```
Comment (teneightfive) [inline assets/stylesheets/components/_progress.scss]: What does `1060px` relate to? Don't think we've used this breakpoint before.

Comment (teneightfive) [inline assets/stylesheets/components/_progress.scss]: Might also be worth looking at on smaller screens. Feels like the center aligned text for first and last items could get cut off at certain resolutions too.

File: assets/stylesheets/components/_progress.scss [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+.c-progress__stage--reset {
+  &::before {
+    background-color: $white;
+    border: 2px solid $grey-3;
+  }
+}
+
+.c-progress__stage--active {
+  @include bold-font(16);
```
Comment (teneightfive) [inline assets/stylesheets/components/_progress.scss]: Not sure about the font size jump for current item. Think bold font weight might be enough.

**Comment by tyom** [comment]
Wonder if the investor company and code should be flipped to be consistent with investment projects pipeline layout.

**Comment by tyom** [comment]
Initial sorting selection seems to have gone
image

File: assets/stylesheets/components/form/_form-control.scss [hunk 5]
```diff
@@ -66,3 +66,11 @@ select.c-form-control {
 .c-form-control--short {
   max-width: 10em;
 }
+
+.c-form-control--medium {
+  max-width: 4em;
+}
+
+.c-form-control--small {
```
Comment (teneightfive) [inline assets/stylesheets/components/form/_form-control.scss]: Maybe we should tweak the current `.c-form-control--short` to be `.c-form-control--long`, add `.c-form-control--medium` and then set `.c-form-control--short` to be `2.6em`?

Avoids us mixing `short` and `small` then.

File: assets/stylesheets/components/form/_form-fieldset.scss [hunk 6]
```diff
@@ -16,6 +17,12 @@
   margin: 0 $default-spacing-unit / 2;
 }
 
+.c-form-fieldset__hint {
```
Comment (teneightfive) [inline assets/stylesheets/components/form/_form-fieldset.scss]: What's this being used for? Couldn't find it in the views.

File: assets/stylesheets/components/form/_form-control.scss (hunk 5 above)
Comment (tyom) [inline assets/stylesheets/components/form/_form-control.scss]: We use `small` and `smaller` to adjust input size.

image

You should pick something different. We already have `short`. Maybe `shorter` and `shortest`?
See above for existing rule `.c-form-control--short`.

File: assets/stylesheets/components/_status-bar.scss [hunk 7]
```diff
@@ -2,39 +2,28 @@
...
+    &:last-child {
+      border-right: 0;
+    }
   }
 }
 
 .status-bar__section-title {
-  font-weight: 700;
+  color: $grey-1;
+  font-size: 16px;
```
Comment (tyom) [inline assets/stylesheets/components/_status-bar.scss]: Apparently we're using `core-font` mixin until we do a review of our typography.

File: src/templates/_components/investment-status.njk [hunk 8]
```diff
@@ -14,28 +13,18 @@
  #}
 
 {% if projectCode %}
-  <div class="status-bar grid-row">
-    <dl class="status-bar__section column-one-quarter">
+  <div class="status-bar l-container">
```
Comment (tyom) [inline src/templates/_components/investment-status.njk]: Wonder if we can reuse `c-meta-item` here which is used on investment collection page and do away with `status-bar` altogether?

File: src/templates/_layouts/datahub-base.njk [hunk 9]
```diff
@@ -9,11 +9,13 @@
 {% from "_macros/_deprecated/trade-elements/textarea.njk" import textarea %}
 {% from "_macros/_deprecated/trade-elements/textbox.njk" import textbox %}
 
-{% from "_macros/form.njk" import EntitySearchForm, Form, MultiStepForm with context %}
+{% from "_macros/form.njk" import EntitySearchForm, Form, InlineForm, MultiStepForm with context %}
 {% from "_macros/form.njk" import Fieldset, TextField, MultipleChoiceField, HiddenField %}
+{% from "_macros/form.njk" import DateField %}
```
Comment (tyom) [inline src/templates/_layouts/datahub-base.njk]: Merge it with the line above?

File: src/templates/_macros/elements.njk [hunk 10]
```diff
@@ -0,0 +1,11 @@
+{##
+ # Render anchor markup
+ # @param {string} props.href - Anchor href
+ # @param {string} props.text - Anchor text
+ # @param {string} [props.target=_self] - Anchor target
+ #}
+{% macro Link(props) %}
```
Comment (tyom) [inline src/templates/_macros/elements.njk]: Not sure how useful this macro is. It's just an HTML element with a couple of attributes that are already restricted to certain values. I think you could justify having it if it had custom classes or wrapper.

File: src/templates/_macros/form.njk [hunk 11]
```diff
@@ -78,6 +78,32 @@
...
 {##
+ # Render simple form with inline button for posting hidden values
+ # @param {object} props - A form object
+ #}
+{% macro InlineForm(props) %}
+  {% set hiddenFields = props.hiddenFields | assign({'_csrf': csrfToken}) %}
+  {% set buttonText = props.buttonText or 'Submit' %}
+
+  {{ ErrorSummary(props.errors) }}
+  <form
```
Comment (tyom) [inline src/templates/_macros/form.njk]: Can you reuse existing `Form` macro here?

File: src/apps/investment-projects/views/create/investment-type.njk [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+      </tr>
+    {% endif %}
+    </tbody>
+  </table>
+
+  {% set investmentOptions = [
+      {
+        value: form.options.investmentTypesObj.fdi.value,
+        label: form.options.investmentTypesObj.fdi.label,
+        hint: Link({
```
Comment (tyom) [inline src/apps/investment-projects/views/create/investment-type.njk]: You could just save that link in a variable and pass it along to macro to avoid creating a new macro.

```nunjucks
{% set investmentTypeFdiLink %}
  <a href="/investment-projects/create/investment-type/info#fdi">Is this an FDI project?</a>
{% endset %}
```

Comment (tyom) [inline src/apps/investment-projects/views/create/investment-type.njk]: Also, what is `info` target? The target is meant to be a frame name. I think you meant `_blank` to open in a new window. Maybe we can do an XHR to open inline? Perhaps a follow-up.

File: assets/stylesheets/components/_progress.scss (hunk 3 above)
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: `1060px` is when they can become centre aligned. The declaration on [line 88](https://github.com/uktrade/data-hub-frontend/pull/395/files/56479d71550351766c224c3c9fe0d6f56c3b7c33#diff-f8bf403629d8d51bfd4ce85105beedd7R88) controls how they behave on smaller screens (pulling them in)

File: assets/stylesheets/components/_progress.scss (hunk 4 above)
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: yeah was thinking about colourblind people and making it apparent enough that something has changed

File: src/templates/_layouts/datahub-base.njk (hunk 9 above)
Comment (feedmypixel) [inline src/templates/_layouts/datahub-base.njk]: was wondering if we need to denote a difference between form elements and form patterns. The DateField felt more "patterny" (for want of a real word) than the others

File: src/templates/_macros/elements.njk (hunk 10 above)
Comment (feedmypixel) [inline src/templates/_macros/elements.njk]: agreed - the reason it was built is the having to send it in to the hint for the radios on the `investment-type` page. I initially just sent in a html string with `| safe` but it felt really hacky. This felt a little cleaner. Open to changing if your good with that.

File: src/templates/_macros/form.njk (hunk 11 above)
Comment (feedmypixel) [inline src/templates/_macros/form.njk]: So differences:
- Error summary outside of form
- Button displayed inside form (not in a form group)

It felt like adding flags to the existing form to handle this was the wrong approach.

File: src/templates/_layouts/datahub-base.njk (hunk 9 above)
Comment (tyom) [inline src/templates/_layouts/datahub-base.njk]: I'm not sure we need that kind of separation at this point. `MultipleChoiceField` is also quite "patterny" but to the end user (developer) it just a component with certain inputs.

File: src/apps/investment-projects/views/create/investment-type.njk (hunk 12 above)
Comment (feedmypixel) [inline src/apps/investment-projects/views/create/investment-type.njk]: `target: 'info'` is to open in new window named `info` so that it doesn't just open different blank pages when clicking different links

Comment (feedmypixel) [inline src/apps/investment-projects/views/create/investment-type.njk]: > You could just save that link in a variable and pass it along to macro to avoid creating a new macro.
This answers the above - will do

File: src/templates/_macros/form.njk (hunk 11 above)
Comment (tyom) [inline src/templates/_macros/form.njk]: Hmm. But you're forking the form here so every time there are global changes to form we must remember to update this one as well. Perhaps we should add a flag to disable `ErrorSummary`. You can already disable form submit button.

File: assets/stylesheets/components/_progress.scss (hunk 4 above)
Comment (teneightfive) [inline assets/stylesheets/components/_progress.scss]: Maybe we need more than colour change on the icon. Could be filled for current state and empty for other states?

**Comment by feedmypixel** [comment]
> Wonder if the investor company and code should be flipped to be consistent with investment projects pipeline layout.

is there a component that can be shared with this?

File: assets/stylesheets/components/form/_form-control.scss (hunk 5 above)
Comment (feedmypixel) [inline assets/stylesheets/components/form/_form-control.scss]: Do 
- `short`
- `shorter`
- `shortest` 

provide enough meaning when they are by themselves?

File: src/templates/_components/investment-status.njk (hunk 8 above)
Comment (feedmypixel) [inline src/templates/_components/investment-status.njk]: sounds like a refactor/alignment for after this PR is in

File: assets/stylesheets/components/_status-bar.scss (hunk 7 above)
Comment (feedmypixel) [inline assets/stylesheets/components/_status-bar.scss]: enforced  👮  🚨

**Comment by feedmypixel** [comment]
> Initial sorting selection seems to have gone in investment pipeline.

@tyom where is this? I can't track it down

File: src/apps/investment-projects/views/_layout.njk [hunk 13]
```diff
@@ -21,6 +25,31 @@
...
+    <p>
+      {% if investmentStatus.currentStage.isComplete %}
+        You have added all required information to complete this stage.
+      {% else %}
+        Complete this stage to move to the {{ investmentStatus.nextStage.name }} stage.
+      {% endif %}
+    </p>
+  {% endset %}
+
+  {% call InlineForm(form | assign({
```
Comment (ztolley) [inline src/apps/investment-projects/views/_layout.njk]: Just for me, can you explain why you use assign?

Comment (ztolley) [inline src/apps/investment-projects/views/_layout.njk]: And could this logic go in a controller instead of a view?

File: src/apps/investment-projects/controllers/create/equity-source.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+      if (searchResult) {
+        searchResult.pagination = buildPagination(req, searchResult)
+      }
+
+      if (company && company.uk_based) {
+        showSearch = true
+      }
+
+      res.locals.company = company
```
Comment (ztolley) [inline src/apps/investment-projects/controllers/create/equity-source.js]: could you use 
```
res.locals = Object.assign({}, res.locals, {
  companyInvestments,
  searchTerm,
```

File: src/apps/investment-projects/controllers/create/project.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  const investmentDetails = req.store.get('investment_details')
+
+  if (investmentDetails) {
+    const investmentType = getMetaDataObj(metadata.investmentTypeOptions.map(transformObjectToOption), investmentDetails.investment_type)
+    const fdiType = getMetaDataObj(metadata.fdiOptions.map(transformObjectToOption), investmentDetails.fdi_type)
+    const nonFdiType = getMetaDataObj(metadata.nonFdiOptions.map(transformObjectToOption), investmentDetails.non_fdi_type)
+
+    res.locals.form = get(res, 'locals.form', { options: {} })
+    res.locals.form.options.investmentDetails = {
+      'investment_type': investmentType,
```
Comment (ztolley) [inline src/apps/investment-projects/controllers/create/project.js]: does it need quotes?

File: src/apps/investment-projects/middleware/forms/details.js [hunk 16]
```diff
@@ -103,7 +111,28 @@ function handleFormPost (req, res, next) {
     })
 }
 
+function validateForm (req, res, next) {
```
Comment (ztolley) [inline src/apps/investment-projects/middleware/forms/details.js]: Should the API be doing validation?

File: src/apps/investment-projects/middleware/index.js [hunk 17]
```diff
@@ -1,9 +1,11 @@
+const collection = require('./collection')
 const forms = require('./forms')
 const shared = require('./shared')
-const collection = require('./collection')
+const team = require('./team')
```
Comment (ztolley) [inline src/apps/investment-projects/middleware/index.js]: huh?

File: src/templates/_macros/form.njk [hunk 18]
```diff
@@ -514,3 +521,43 @@
     {% endfor %}
   {% endif %}
 {% endmacro %}
+
```
Comment (ztolley) [inline src/templates/_macros/form.njk]: :)

File: assets/stylesheets/components/form/_form-group.scss [hunk 19]
```diff
@@ -4,7 +4,8 @@
   max-width: 40em;
 
   [type="hidden"] + &,
-  .c-form-fieldset__legend + & {
+  .c-form-fieldset__legend + &,
```
Comment (ztolley) [inline assets/stylesheets/components/form/_form-group.scss]: Some of this stuff should be split into a separate PR maybe as it affects things globally but then again I know that allot of effort for the same result so no worries.

File: src/apps/components/views/form.njk [hunk 20]
```diff
@@ -180,6 +180,19 @@
...
       {% endcall %}
+
+      {{ DateFieldset({
+        name: 'date',
+        label: 'What is your favourite day?',
+        hint: 'A day you really like',
+        error: form.errors.messages.date,
+        inputName: 'date',
+        inputValue: {
+          year: form.state.dateYear,
```
Comment (ztolley) [inline src/apps/components/views/form.njk]: I wonder if you could have a field named date with sub properties or date in a set format and split with filters or just pass data in a set format and the control splits it.

How does it handle multiple dates in a form?

**Comment by ztolley** [comment]
In a way this PR is actually 5 or 6 PR's in one. Things like the session storage, the steps support and the date control. But all good stuff.

**Comment by feedmypixel** [comment]
> Just for me, can you explain why you use assign?

because I am merging options into the [form](https://github.com/uktrade/data-hub-frontend/pull/395/files#diff-11b25278c870c6bb98a48ece404c54a8R42). The form object can potentially hold error state. It feels more appropriate to have the properties in the object that is being assigned in the view layer.

**Comment by feedmypixel** [comment]
> And could this logic go in a controller instead of a view?

Explained above

File: src/apps/investment-projects/middleware/index.js (hunk 17 above)
Comment (feedmypixel) [inline src/apps/investment-projects/middleware/index.js]: alphabeticalness

File: src/templates/_macros/form.njk [hunk 21]
```diff
@@ -488,3 +527,43 @@
...
+{% macro DateFieldset(props) %}
+  {% set fieldId = 'field-' + props.name if props.name %}
+
+  {% call FormGroup(props | assign({ fieldId: fieldId, element: 'fieldset', modifier: 'inline' })) %}
+    {{ TextField({
+      modifier: ['inline', 'shortest'],
+      name: props.inputName + '_day',
+      value: props.inputValue.day,
+      label: 'Day',
+      hint: 'dd'
```
Comment (teneightfive) [inline src/templates/_macros/form.njk]: I would drop these hints from this component. We may well decide to support different formats, in fact, I think we should. We shouldn't require a use to enter `01` for the first of the month.

File: src/templates/_macros/form.njk [hunk 22]
```diff
@@ -488,3 +527,43 @@
...
+ #}
+{% macro DateFieldset(props) %}
+  {% set fieldId = 'field-' + props.name if props.name %}
+
+  {% call FormGroup(props | assign({ fieldId: fieldId, element: 'fieldset', modifier: 'inline' })) %}
+    {{ TextField({
+      modifier: ['inline', 'shortest'],
+      name: props.inputName + '_day',
+      value: props.inputValue.day,
+      label: 'Day',
```
Comment (teneightfive) [inline src/templates/_macros/form.njk]: We also talked about removing the bold from these labels so may need a modifier somewhere.

File: src/templates/_macros/form.njk (hunk 11 above)
Comment (feedmypixel) [inline src/templates/_macros/form.njk]: So the InlineForm macro has now gone, after a rethink there is no need for this macro

File: src/apps/investment-projects/middleware/forms/details.js (hunk 16 above)
Comment (feedmypixel) [inline src/apps/investment-projects/middleware/forms/details.js]: This is validating the UI radio buttons, these button values are not sent to the API. They are just a UI pattern to make the entering `client_relationship_manager` and `referral_source_adviser` easier

File: src/apps/components/views/form.njk (hunk 20 above)
Comment (feedmypixel) [inline src/apps/components/views/form.njk]: > How does it handle multiple dates in a form?

Its encapsulated in the macro. This is the `/components` example a better of example of how this can be handled can be seen [here](https://github.com/uktrade/data-hub-frontend/pull/395/commits/8147aea82d866c1a0751de1d9d90c5b6dfd968b8#diff-f6a833c01c454216b27811d03a17b2c4R304)

File: assets/stylesheets/components/_local-header.scss [hunk 23]
```diff
@@ -14,18 +14,18 @@
...
-    margin-bottom: -$default-spacing-unit;
-  }
-
   .c-entity-search__aggregations {
     margin-bottom: -$default-spacing-unit * 2;
   }
+
+  .c-progress__stage {
+    &::before {
+      outline-color: $grey-4;
```
Comment (feedmypixel) [inline assets/stylesheets/components/_local-header.scss]: This is needed as the progress is either within the header in `investment-projects` or on a white background in `/components`. It felt like adding it here and defaulting to white rather than providing a modifier was the cleaner way of doing it. This controls the gap around the circles and the progress line so `transparent won't work`

screen shot 2017-08-03 at 14 41 49

**Comment by tyom** [comment]
> > Wonder if the investor company and code should be flipped to be consistent with investment projects pipeline layout.
>
> is there a component that can be shared with this?

No, there's no component. But we could match it stylistically. But that's something we could look at later on.

File: assets/stylesheets/components/_progress.scss [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+    z-index: 20;
+  }
+
+  &.c-progress__bar--0-percent {
+    &::after {
+      width: 0;
+    }
+  }
+
+  &.c-progress__bar--25--percent {
```
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: >  tries to get me to avoid modifiers on elements if possible

👍

File: assets/stylesheets/components/_progress.scss [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+  &::before {
+    position: relative;
+    width: 14px;
+    height: 14px;
+    margin: -7px -7px 0;
+    content: "";
+    border-radius: 20px;
+    border: 2px solid $govuk-blue;
+    background-color: $govuk-blue;
+    outline: 2px solid $white;
```
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: outline is providing the gap between the circle and the line. Mentioned above `transparent` won't work here.

screen shot 2017-08-03 at 14 41 49

File: assets/stylesheets/components/_progress.scss [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+  }
+}
+
+.c-progress__stage-title {
+  position: relative;
+  left: -50%;
+}
+
+.c-progress__bar {
+  > li {
```
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: with refactor these declarations have now gone

File: assets/stylesheets/components/_progress.scss [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+      left: -85%;
+    }
+  }
+}
+
+@include media($min-width: 1060px) {
+  .c-progress__bar {
+    width: 100%;
+
+    > li {
```
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: with refactor these declarations have now gone

File: assets/stylesheets/components/_progress.scss [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+      &:last-child {
+        .c-progress__stage-title {
+          left: -50%;
+        }
+      }
+    }
+  }
+}
+
+.c-progress__stage--part-1 {
```
Comment (feedmypixel) [inline assets/stylesheets/components/_progress.scss]: I have avoided `:nth-child` so this work covers all browsers. I have updated the names to `c-progress__stage--one`, `c-progress__stage--two`, etc... This work is inside of a for loop so unlessI make `c-progress__stage` a block it would be tricky to add it to the block

File: assets/stylesheets/components/_status-bar.scss [hunk 29]
```diff
@@ -2,39 +2,28 @@
 @import "../settings/layout";
 @import "../tools/mixins";
 
-.status-bar {
```
Comment (feedmypixel) [inline assets/stylesheets/components/_status-bar.scss]: 👍  have updated the markup and BEM names as `status-bar` styles have changend

File: assets/stylesheets/components/_progress.scss (hunk 24 above)
Comment (tyom) [inline assets/stylesheets/components/_progress.scss]: I feel like the modifier should be applied to the `c-progress` and not to specifically to `__bar`. The modifier may affect other elements within the component. Say you wanted to turn to text colour to green when it's complete or add additional pseudo element. Conceptually the percentage can be applied to the whole progress not just the gauge measuring it.

File: src/apps/investment-projects/controllers/create/equity-source.js [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+
+function renderEquitySourcePage (req, res, next) {
+  return res
+    .breadcrumb('Add investment project')
+    .render('investment-projects/views/create/equity-source')
+}
+
+function getHandler (req, res, next) {
+  const company = res.locals.company
+  const searchTerm = req.query['q']
```
Comment (feedmypixel) [inline src/apps/investment-projects/controllers/create/equity-source.js]: this was previously in `create-1`, will change to `term` to be the same as the home page search

File: src/apps/components/views/progress.njk [hunk 31]
```diff
@@ -0,0 +1,11 @@
+{% extends "./_layout.njk" %}
+
+{% block body_main_content %}
+  <h2 class="page-heading">Progress Indicator</h2>
```
Comment (feedmypixel) [inline src/apps/components/views/progress.njk]: yeah this really became apparent when adding to `/components`

File: src/apps/investment-projects/controllers/create/equity-source.js [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+      res.locals.showSearch = showSearch
+
+      next()
+    })
+    .catch(next)
+}
+
+function postHandler (req, res, next) {
+  const isEquitySource = req.body['is_equity_source']
+  const clientCompanyId = req.body['company_id']
```
Comment (feedmypixel) [inline src/apps/investment-projects/controllers/create/equity-source.js]: yup - this page feels like it was doing a bit much in the first place.
- Radios for is equity source,
- search for equity search,
- search results for equity source

 But I decided to keep the changes to a minimum from the previous `create-1`. Once I understand more about how we are approaching flows next week I will look at this again and potentially split out.

File: src/apps/investment-projects/controllers/create/investment-type.js [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+    'fdi_type': req.body.fdi_type,
+    'non_fdi_type': req.body.non_fdi_type,
+  })
+
+  if (
+    investmentTypeId === investmentTypeOptions.fdi.value ||
+    investmentTypeId === investmentTypeOptions.non_fdi.value
+  ) {
+    return res.redirect(`/investment-projects/create/equity-source/${companyId}`)
+  } else {
```
Comment (feedmypixel) [inline src/apps/investment-projects/controllers/create/investment-type.js]: yup - its error reporting if there has been no selection of a radio button

Comment (feedmypixel) [inline src/apps/investment-projects/controllers/create/investment-type.js]: have rejigged. This is clearer now. Previously a symptom of the Commitment to invest flow not being done yet

File: src/apps/investment-projects/middleware/forms/details.js [hunk 34]
```diff
@@ -93,8 +99,10 @@ function handleFormPost (req, res, next) {
     .catch((err) => {
       if (err.statusCode === 400) {
         res.locals.form = get(res, 'locals.form', {})
```
Comment (feedmypixel) [inline src/apps/investment-projects/middleware/forms/details.js]: 👍  if `res.locals.form` is `undefined` `Object.assign` handles it appropriately

File: src/middleware/form.js [hunk 35]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Simple util to send form object when using a res.redirect. Mimics res.locals functionality
+ * @returns {form}
+ */
+module.exports = () => {
+  return function form (req, res, next) {
+    req.session.form = req.session.form || []
```
Comment (feedmypixel) [inline src/middleware/form.js]: this has gone now but no form was an `Array` which you pushed and popped form objects from onto `locals`

File: src/apps/transformers.js [hunk 36]
```diff
@@ -21,7 +22,35 @@ function transformContactToOption ({ id, first_name, last_name }) {
   }
 }
 
+/**
+ * Utility to build an object from a transformed metadata array of objects so you can reference properties
+ * by key rather than array index. Helpful when the array length changes.
+ * @returns {{}}
+ */
+function buildMetaDataObj (collection) {
```
Comment (feedmypixel) [inline src/apps/transformers.js]: `keyBy` has done the trick

File: src/middleware/form.js [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+
+    req.form = (errors) => {
```
Comment (feedmypixel) [inline src/middleware/form.js]: this functionality has gone now

File: src/apps/investment-projects/views/_details-form.njk [hunk 38]
```diff
@@ -1,170 +1,299 @@
...
+
+{% if investmentData.id %}
+  <div class="c-form-group">
+    {{ MultipleChoiceField({
+      type: 'radio',
+      name: 'investment_type',
+      label: 'Type of investment',
+      value: form.state['investment_type'],
+      error: form.errors.messages['investment_type'],
+      children:  [
```
Comment (tyom) [inline src/apps/investment-projects/views/_details-form.njk]: You don't need `children`. You can just call that macro and nest stuff in it. We should avoid multiple ways of doing the same thing.

File: src/templates/_macros/form.njk [hunk 39]
```diff
@@ -429,6 +438,10 @@
       </option>
     {% endfor %}
   </select>
+
+  {% for child in fieldChildren %}
```
Comment (tyom) [inline src/templates/_macros/form.njk]: I'm not convinced this is needed.

File: src/apps/investment-projects/views/_details-form.njk (hunk 38 above)
Comment (tyom) [inline src/apps/investment-projects/views/_details-form.njk]: [An example of call](https://github.com/uktrade/data-hub-frontend/blob/develop/src/apps/investment-projects/views/value-edit.njk#L12-L36).

Comment (feedmypixel) [inline src/apps/investment-projects/views/_details-form.njk]: > You don't need children. 

do you mean use [caller](https://github.com/uktrade/data-hub-frontend/blob/dcc35618e8c93e603a714e1ef346a8557074c098/src/templates/_macros/form.njk#L329).

> We should avoid multiple ways of doing the same thing.

This `select` pattern is the [same pattern](https://github.com/uktrade/data-hub-frontend/blob/dcc35618e8c93e603a714e1ef346a8557074c098/src/templates/_macros/form.njk#L493-L495) used for `radio` and `checkbox` when using the [MultipleChoiceField macro](https://github.com/uktrade/data-hub-frontend/blob/dcc35618e8c93e603a714e1ef346a8557074c098/src/templates/_macros/form.njk#L320)

Comment (feedmypixel) [inline src/apps/investment-projects/views/_details-form.njk]: Use case is on the create form

business

Comment (feedmypixel) [inline src/apps/investment-projects/views/_details-form.njk]: will align this work to use `caller` rather than children

File: assets/stylesheets/components/_meta-list.scss [hunk 40]
```diff
@@ -0,0 +1,11 @@
+@import "../settings";
+
+.c-meta-list {
+  .c-meta-item {
+    margin-right: $default-spacing-unit;
```
Comment (teneightfive) [inline assets/stylesheets/components/_meta-list.scss]: A left margin when there is a sibling might be better here. But could be addressed later during a spacing review.

```scss
.c-meta-item + .c-meta-item {
  margin-left: $default-spacing-unit;
}
```
